Add `__auto_type` support

> you may also declare the type of a variable as __auto_type. In that case, the declaration must declare only one variable, whose declarator must just be an identifier, the declaration must be initialized, and the type of the variable is determined by the initializer; the name of the variable is not in scope until after the initializer.

Examples of errors:
```
test.c:6:17: error: declaration of variable 'x' with deduced type '__auto_type' requires an initializer
    __auto_type x;
                ^
1 error generated.
➜  ~ gcc test.c
test.c:6:21: error: cannot use __auto_type with initializer list in C
    __auto_type x = {1,2};
                    ^
1 error generated.
➜  ~ gcc test.c
test.c:6:18: error: 'x' declared as array of '__auto_type'
    __auto_type x[2] = {1,2};
                 ^
1 error generated.
```

The deduced type when a string literal initializer is used is `char *`

See bottom of this page: https://gcc.gnu.org/onlinedocs/gcc/Typeof.html

--- a/src/Attribute.zig
+++ b/src/Attribute.zig
@@ -1277,12 +1277,13 @@ pub fn applyEnumeratorAttributes(p: *Parser, ty: Type, attr_buf_start: usize) !T
 
 fn applyAligned(attr: Attribute, p: *Parser, ty: Type, tag: ?Diagnostics.Tag) !void {
     const base = ty.canonicalize(.standard);
-    const default_align = base.alignof(p.comp);
     if (attr.args.aligned.alignment) |alignment| alignas: {
         if (attr.syntax != .keyword) break :alignas;
 
         const align_tok = attr.args.aligned.__name_tok;
         if (tag) |t| try p.errTok(t, align_tok);
+
+        const default_align = base.alignof(p.comp);
         if (ty.isFunc()) {
             try p.errTok(.alignas_on_func, align_tok);
         } else if (alignment.requested < default_align) {

--- a/src/Diagnostics.zig
+++ b/src/Diagnostics.zig
@@ -160,6 +160,7 @@ pub const Options = packed struct {
     @"pre-c2x-compat": Kind = .default,
     @"pointer-bool-conversion": Kind = .default,
     @"string-conversion": Kind = .default,
+    @"gnu-auto-type": Kind = .default,
 };
 
 const messages = struct {
@@ -2240,6 +2241,47 @@ const messages = struct {
         const opt = "c2x-extensions";
         const kind = .warning;
         const suppress_version = .c2x;
+    };
+    const auto_type_extension = struct {
+        const msg = "'__auto_type' is a GNU extension";
+        const opt = "gnu-auto-type";
+        const kind = .off;
+        const pedantic = true;
+    };
+    const auto_type_not_allowed = struct {
+        const msg = "'__auto_type' not allowed in {s}";
+        const kind = .@"error";
+        const extra = .str;
+    };
+    const auto_type_requires_initializer = struct {
+        const msg = "declaration of variable '{s}' with deduced type requires an initializer";
+        const kind = .@"error";
+        const extra = .str;
+    };
+    const auto_type_requires_single_declarator = struct {
+        const msg = "'__auto_type' may only be used with a single declarator";
+        const kind = .@"error";
+    };
+    const auto_type_requires_plain_declarator = struct {
+        const msg = "'__auto_type' requires a plain identifier as declarator";
+        const kind = .@"error";
+    };
+    const invalid_cast_to_auto_type = struct {
+        const msg = "Invalid cast to '__auto_type'";
+        const kind = .@"error";
+    };
+    const auto_type_from_bitfield = struct {
+        const msg = "cannot use bit-field as '__auto_type' initializer";
+        const kind = .@"error";
+    };
+    const array_of_auto_type = struct {
+        const msg = "'{s}' declared as array of '__auto_type'";
+        const kind = .@"error";
+        const extra = .str;
+    };
+    const auto_type_with_init_list = struct {
+        const msg = "cannot use '__auto_type' with initializer list";
+        const kind = .@"error";
     };
 };
 

--- a/src/Parser.zig
+++ b/src/Parser.zig
@@ -821,6 +821,11 @@ fn decl(p: *Parser) Error!bool {
 
     // Check for function definition.
     if (init_d.d.func_declarator != null and init_d.initializer.node == .none and init_d.d.ty.isFunc()) fn_def: {
+        if (decl_spec.auto_type) |tok_i| {
+            try p.errStr(.auto_type_not_allowed, tok_i, "function return type");
+            return error.ParsingFailed;
+        }
+
         switch (p.tok_ids[p.tok_i]) {
             .comma, .semicolon => break :fn_def,
             .l_brace => {},
@@ -921,7 +926,7 @@ fn decl(p: *Parser) Error!bool {
         } else {
             for (init_d.d.ty.params()) |param| {
                 if (param.ty.hasUnboundVLA()) try p.errTok(.unbound_vla, param.name_tok);
-                if (param.ty.hasIncompleteSize() and !param.ty.is(.void)) try p.errStr(.parameter_incomplete_ty, param.name_tok, try p.typeStr(param.ty));
+                if (param.ty.hasIncompleteSize() and !param.ty.is(.void) and param.ty.specifier != .invalid) try p.errStr(.parameter_incomplete_ty, param.name_tok, try p.typeStr(param.ty));
 
                 if (param.name == .empty) {
                     try p.errTok(.omitting_parameter_name, param.name_tok);
@@ -1000,6 +1005,8 @@ fn decl(p: *Parser) Error!bool {
         }
 
         if (p.eatToken(.comma) == null) break;
+
+        if (decl_spec.auto_type) |tok_i| try p.errTok(.auto_type_requires_single_declarator, tok_i);
 
         init_d = (try p.initDeclarator(&decl_spec, attr_buf_top)) orelse {
             try p.err(.expected_ident_or_l_paren);
@@ -1127,6 +1134,7 @@ pub const DeclSpec = struct {
     constexpr: ?TokenIndex = null,
     @"inline": ?TokenIndex = null,
     noreturn: ?TokenIndex = null,
+    auto_type: ?TokenIndex = null,
     ty: Type,
 
     fn validateParam(d: DeclSpec, p: *Parser, ty: *Type) Error!void {
@@ -1139,6 +1147,10 @@ pub const DeclSpec = struct {
         if (d.@"inline") |tok_i| try p.errStr(.func_spec_non_func, tok_i, "inline");
         if (d.noreturn) |tok_i| try p.errStr(.func_spec_non_func, tok_i, "_Noreturn");
         if (d.constexpr) |tok_i| try p.errTok(.invalid_storage_on_param, tok_i);
+        if (d.auto_type) |tok_i| {
+            try p.errStr(.auto_type_not_allowed, tok_i, "function prototype");
+            ty.* = Type.invalid;
+        }
     }
 
     fn validateFnDef(d: DeclSpec, p: *Parser) Error!Tree.Tag {
@@ -1347,6 +1359,7 @@ fn declSpec(p: *Parser) Error!?DeclSpec {
     if (p.tok_i == start) return null;
 
     d.ty = try spec.finish(p);
+    d.auto_type = spec.auto_type_tok;
     return d;
 }
 
@@ -1543,8 +1556,11 @@ fn initDeclarator(p: *Parser, decl_spec: *DeclSpec, attr_buf_top: usize) Error!?
     try p.attributeSpecifierExtra(init_d.d.name);
 
     var apply_var_attributes = false;
-
     if (decl_spec.storage_class == .typedef) {
+        if (decl_spec.auto_type) |tok_i| {
+            try p.errStr(.auto_type_not_allowed, tok_i, "typedef");
+            return error.ParsingFailed;
+        }
         init_d.d.ty = try Attribute.applyTypeAttributes(p, init_d.d.ty, attr_buf_top, null);
     } else if (init_d.d.ty.isFunc()) {
         init_d.d.ty = try Attribute.applyFunctionAttributes(p, init_d.d.ty, attr_buf_top);
@@ -1581,10 +1597,21 @@ fn initDeclarator(p: *Parser, decl_spec: *DeclSpec, attr_buf_top: usize) Error!?
             init_d.d.ty.specifier = .array;
         }
     }
+
+    const name = init_d.d.name;
+    if (init_d.d.ty.is(.auto_type)) {
+        if (init_d.initializer.node == .none) {
+            init_d.d.ty = Type.invalid;
+            try p.errStr(.auto_type_requires_initializer, name, p.tokSlice(name));
+            return init_d;
+        } else {
+            init_d.d.ty.specifier = init_d.initializer.ty.specifier;
+            init_d.d.ty.data = init_d.initializer.ty.data;
+        }
+    }
     if (apply_var_attributes) {
         init_d.d.ty = try Attribute.applyVariableAttributes(p, init_d.d.ty, attr_buf_top, null);
     }
-    const name = init_d.d.name;
     if (decl_spec.storage_class != .typedef and init_d.d.ty.hasIncompleteSize()) incomplete: {
         const specifier = init_d.d.ty.canonicalize(.standard).specifier;
         if (decl_spec.storage_class == .@"extern") switch (specifier) {
@@ -1610,6 +1637,7 @@ fn initDeclarator(p: *Parser, decl_spec: *DeclSpec, attr_buf_top: usize) Error!?
 
 /// typeSpec
 ///  : keyword_void
+///  | keyword_auto_type
 ///  | keyword_char
 ///  | keyword_short
 ///  | keyword_int
@@ -1645,6 +1673,10 @@ fn typeSpec(p: *Parser, ty: *Type.Builder) Error!bool {
         if (try p.typeQual(&ty.qual)) continue;
         switch (p.tok_ids[p.tok_i]) {
             .keyword_void => try ty.combine(p, .void, p.tok_i),
+            .keyword_auto_type => {
+                try p.errTok(.auto_type_extension, p.tok_i);
+                try ty.combine(p, .auto_type, p.tok_i);
+            },
             .keyword_bool, .keyword_c23_bool => try ty.combine(p, .bool, p.tok_i),
             .keyword_int8, .keyword_int8_2, .keyword_char => try ty.combine(p, .char, p.tok_i),
             .keyword_int16, .keyword_int16_2, .keyword_short => try ty.combine(p, .short, p.tok_i),
@@ -2038,6 +2070,10 @@ fn recordDeclarator(p: *Parser) Error!bool {
         // 0 means unnamed
         var name_tok: TokenIndex = 0;
         var ty = base_ty;
+        if (ty.is(.auto_type)) {
+            try p.errStr(.auto_type_not_allowed, p.tok_i, if (p.record.kind == .keyword_struct) "struct member" else "union member");
+            ty = Type.invalid;
+        }
         var bits_node: NodeIndex = .none;
         var bits: ?u32 = null;
         const first_tok = p.tok_i;
@@ -2144,7 +2180,7 @@ fn recordDeclarator(p: *Parser) Error!bool {
                 }
             }
             p.record.flexible_field = first_tok;
-        } else if (ty.hasIncompleteSize()) {
+        } else if (ty.specifier != .invalid and ty.hasIncompleteSize()) {
             try p.errStr(.field_incomplete_ty, first_tok, try p.typeStr(ty));
         } else if (p.record.flexible_field) |some| {
             if (some != first_tok and p.record.kind == .keyword_struct) try p.errTok(.flexible_non_final, some);
@@ -2597,6 +2633,10 @@ fn declarator(
 ) Error!?Declarator {
     const start = p.tok_i;
     var d = Declarator{ .name = 0, .ty = try p.pointer(base_type) };
+    if (base_type.is(.auto_type) and !d.ty.is(.auto_type)) {
+        try p.errTok(.auto_type_requires_plain_declarator, start);
+        return error.ParsingFailed;
+    }
 
     const maybe_ident = p.tok_i;
     if (kind != .abstract and (try p.eatIdentifier()) != null) {
@@ -2695,6 +2735,10 @@ fn directDeclarator(p: *Parser, base_type: Type, d: *Declarator, kind: Declarato
         }
         if (static) |_| try size.expect(p);
 
+        if (base_type.is(.auto_type)) {
+            try p.errStr(.array_of_auto_type, d.name, p.tokSlice(d.name));
+            return error.ParsingFailed;
+        }
         const outer = try p.directDeclarator(base_type, d, kind);
         var max_bits = p.comp.target.ptrBitWidth();
         if (max_bits > 61) max_bits = 61;
@@ -2932,6 +2976,10 @@ fn initializer(p: *Parser, init_ty: Type) Error!Result {
         if (try p.coerceArrayInit(&res, tok, init_ty)) return res;
         try p.coerceInit(&res, tok, init_ty);
         return res;
+    }
+    if (init_ty.is(.auto_type)) {
+        try p.err(.auto_type_with_init_list);
+        return error.ParsingFailed;
     }
 
     var il: InitList = .{};
@@ -3347,7 +3395,15 @@ fn coerceArrayInit(p: *Parser, item: *Result, tok: TokenIndex, target: Type) !bo
 fn coerceInit(p: *Parser, item: *Result, tok: TokenIndex, target: Type) !void {
     if (target.is(.void)) return; // Do not do type coercion on excess items
 
+    const node = item.node;
     try item.lvalConversion(p);
+    if (target.is(.auto_type)) {
+        if (p.getNode(node, .member_access_expr) orelse p.getNode(node, .member_access_ptr_expr)) |member_node| {
+            if (Tree.isBitfield(p.nodes.slice(), member_node)) try p.errTok(.auto_type_from_bitfield, tok);
+        }
+        return;
+    }
+
     try item.coerce(p, target, tok, .init);
 }
 
@@ -5152,7 +5208,11 @@ const Result = struct {
                 res.val.intCast(res.ty, to, p.comp);
             }
         } else {
-            try p.errStr(.invalid_cast_type, tok, try p.typeStr(to));
+            if (to.is(.auto_type)) {
+                try p.errTok(.invalid_cast_to_auto_type, tok);
+            } else {
+                try p.errStr(.invalid_cast_type, tok, try p.typeStr(to));
+            }
             return error.ParsingFailed;
         }
         if (to.anyQual()) try p.errStr(.qual_cast, tok, try p.typeStr(to));

--- a/src/Tokenizer.zig
+++ b/src/Tokenizer.zig
@@ -139,6 +139,7 @@ pub const Token = struct {
         macro_pretty_func,
 
         keyword_auto,
+        keyword_auto_type,
         keyword_break,
         keyword_case,
         keyword_char,
@@ -313,6 +314,7 @@ pub const Token = struct {
                 .macro_function,
                 .macro_pretty_func,
                 .keyword_auto,
+                .keyword_auto_type,
                 .keyword_break,
                 .keyword_case,
                 .keyword_char,
@@ -555,6 +557,7 @@ pub const Token = struct {
                 .hash_hash => "##",
 
                 .keyword_auto => "auto",
+                .keyword_auto_type => "__auto_type",
                 .keyword_break => "break",
                 .keyword_case => "case",
                 .keyword_char => "char",
@@ -908,6 +911,7 @@ pub const Token = struct {
         .{ "__PRETTY_FUNCTION__", .macro_pretty_func },
 
         // gcc keywords
+        .{ "__auto_type", .keyword_auto_type },
         .{ "__const", .keyword_const1 },
         .{ "__const__", .keyword_const2 },
         .{ "__inline", .keyword_inline1 },
@@ -1819,7 +1823,7 @@ test "operators" {
 
 test "keywords" {
     try expectTokens(
-        \\auto break case char const continue default do 
+        \\auto __auto_type break case char const continue default do
         \\double else enum extern float for goto if int 
         \\long register return short signed sizeof static 
         \\struct switch typedef union unsigned void volatile 
@@ -1829,6 +1833,7 @@ test "keywords" {
         \\
     , &.{
         .keyword_auto,
+        .keyword_auto_type,
         .keyword_break,
         .keyword_case,
         .keyword_char,

--- a/src/Type.zig
+++ b/src/Type.zig
@@ -273,6 +273,10 @@ pub const Specifier = enum {
     /// A NaN-like poison value
     invalid,
 
+    /// GNU auto type
+    /// This is a placeholder specifier - it must be replaced by the actual type specifier (determined by the initializer)
+    auto_type,
+
     void,
     bool,
 
@@ -917,6 +921,7 @@ pub fn sizeCompare(a: Type, b: Type, comp: *Compilation) TypeSizeOrder {
 /// Size of type as reported by sizeof
 pub fn sizeof(ty: Type, comp: *const Compilation) ?u64 {
     return switch (ty.specifier) {
+        .auto_type => unreachable,
         .variable_len_array, .unspecified_variable_len_array => return null,
         .incomplete_array => return if (comp.langopts.emulate == .msvc) @as(?u64, 0) else null,
         .func, .var_args_func, .old_style_func, .void, .bool => 1,
@@ -1019,6 +1024,7 @@ pub fn alignof(ty: Type, comp: *const Compilation) u29 {
 
     return switch (ty.specifier) {
         .invalid => unreachable,
+        .auto_type => unreachable,
 
         .variable_len_array,
         .incomplete_array,
@@ -1182,6 +1188,7 @@ pub fn eql(a_param: Type, b_param: Type, comp: *const Compilation, check_qualifi
     const a = a_param.canonicalize(.standard);
     const b = b_param.canonicalize(.standard);
 
+    if (a.specifier == .invalid or b.specifier == .invalid) return false;
     if (a.alignof(comp) != b.alignof(comp)) return false;
     if (a.isPtr()) {
         if (!b.isPtr()) return false;
@@ -1418,6 +1425,7 @@ pub fn validateCombinedType(ty: Type, p: *Parser, source_tok: TokenIndex) Parser
 pub const Builder = struct {
     complex_tok: ?TokenIndex = null,
     bit_int_tok: ?TokenIndex = null,
+    auto_type_tok: ?TokenIndex = null,
     typedef: ?struct {
         tok: TokenIndex,
         ty: Type,
@@ -1432,6 +1440,8 @@ pub const Builder = struct {
     pub const Specifier = union(enum) {
         none,
         void,
+        /// GNU __auto_type extension
+        auto_type,
         nullptr_t,
         bool,
         char,
@@ -1543,6 +1553,7 @@ pub const Builder = struct {
             return switch (spec) {
                 .none => unreachable,
                 .void => "void",
+                .auto_type => "__auto_type",
                 .nullptr_t => "nullptr_t",
                 .bool => if (langopts.standard.atLeast(.c2x)) "bool" else "_Bool",
                 .char => "char",
@@ -1677,6 +1688,7 @@ pub const Builder = struct {
                 }
             },
             .void => ty.specifier = .void,
+            .auto_type => ty.specifier = .auto_type,
             .nullptr_t => unreachable, // nullptr_t can only be accessed via typeof(nullptr)
             .bool => ty.specifier = .bool,
             .char => ty.specifier = .char,
@@ -1909,8 +1921,12 @@ pub const Builder = struct {
             try p.errStr(.invalid_typeof, source_tok, @tagName(new));
         }
 
-        if (new == .complex) b.complex_tok = source_tok;
-        if (new == .bit_int) b.bit_int_tok = source_tok;
+        switch (new) {
+            .complex => b.complex_tok = source_tok,
+            .bit_int => b.bit_int_tok = source_tok,
+            .auto_type => b.auto_type_tok = source_tok,
+            else => {},
+        }
 
         if (new == .int128 and !target_util.hasInt128(p.comp.target)) {
             try p.errStr(.type_not_supported_on_target, source_tok, "__int128");
@@ -2097,6 +2113,10 @@ pub const Builder = struct {
                 .complex_unsigned => .{ .complex_ubit_int = new.bit_int },
                 else => return b.cannotCombine(p, source_tok),
             },
+            .auto_type => b.specifier = switch (b.specifier) {
+                .none => .auto_type,
+                else => return b.cannotCombine(p, source_tok),
+            },
             .fp16 => b.specifier = switch (b.specifier) {
                 .none => .fp16,
                 else => return b.cannotCombine(p, source_tok),
@@ -2213,6 +2233,7 @@ pub const Builder = struct {
     pub fn fromType(ty: Type) Builder.Specifier {
         return switch (ty.specifier) {
             .void => .void,
+            .auto_type => .auto_type,
             .nullptr_t => .nullptr_t,
             .bool => .bool,
             .char => .char,

--- a/test/cases/__auto_type.c
+++ b/test/cases/__auto_type.c
@@ -1,0 +1,82 @@
+//aro-args -Wno-gnu-alignof-expression
+#include "test_helpers.h"
+__auto_type foo(void);
+__auto_type foo1(void) {}
+
+void bar(__auto_type);
+void bar1(__auto_type x) {}
+
+typedef __auto_type my_auto_type;
+int __auto_type a = 5;
+
+struct S {
+    unsigned bitfield: 5;
+};
+struct BadStruct {
+    __auto_type y;
+};
+union BadUnion {
+    __auto_type x;
+};
+
+int myfunc(int x) { return x; }
+
+void baz(void) {
+
+#   pragma GCC diagnostic push
+#   pragma GCC diagnostic warning "-Wgnu-auto-type"
+#   pragma GCC diagnostic ignored "-Wsizeof-array-argument"
+    __auto_type decayed_arr = (double[]){1.,2.,3.,4.,5.};
+    _Static_assert(sizeof(decayed_arr) == sizeof(double *), "");
+#   pragma GCC diagnostic pop
+
+    __auto_type b = 5U;
+    EXPECT_TYPE(b, unsigned);
+    int c = (__auto_type)5;
+
+    __auto_type d;
+
+    __auto_type e = 1, f = 2;
+
+    struct S s = {};
+    __auto_type g = s.bitfield;
+
+    const int const_int = 5;
+    __auto_type h = const_int;
+    h = 100;
+
+    const __auto_type i = 0;
+    i += 1;
+
+    __auto_type func_ptr = myfunc;
+    EXPECT_TYPE(func_ptr, __typeof__(&myfunc));
+    EXPECT_TYPE(func_ptr, int (*)(int));
+
+    __auto_type my_struct = (struct S){.bitfield = 10 };
+    EXPECT_TYPE(my_struct, struct S);
+
+    __auto_type auto_array[2] = {1, 2};
+
+    __auto_type init_list = {1,2};
+
+    __attribute__((aligned(128))) __auto_type aligned_var = 0ULL;
+    _Static_assert(_Alignof(aligned_var) == 128, "");
+}
+
+#define EXPECTED_ERRORS "__auto_type.c:3:1: error: '__auto_type' not allowed in function return type" \
+    "__auto_type.c:4:1: error: '__auto_type' not allowed in function return type" \
+    "__auto_type.c:6:10: error: '__auto_type' not allowed in function prototype" \
+    "__auto_type.c:7:11: error: '__auto_type' not allowed in function prototype" \
+    "__auto_type.c:9:9: error: '__auto_type' not allowed in typedef" \
+    "__auto_type.c:10:5: error: cannot combine with previous 'int' specifier" \
+    "__auto_type.c:16:17: error: '__auto_type' not allowed in struct member" \
+    "__auto_type.c:19:17: error: '__auto_type' not allowed in union member" \
+    "__auto_type.c:29:5: warning: '__auto_type' is a GNU extension [-Wgnu-auto-type]" \
+    "__auto_type.c:35:13: error: Invalid cast to '__auto_type'" \
+    "__auto_type.c:37:17: error: declaration of variable 'd' with deduced type requires an initializer" \
+    "__auto_type.c:39:5: error: '__auto_type' may only be used with a single declarator" \
+    "__auto_type.c:42:21: error: cannot use bit-field as '__auto_type' initializer" \
+    "__auto_type.c:49:7: error: expression is not assignable" \
+    "__auto_type.c:58:17: error: 'auto_array' declared as array of '__auto_type'" \
+    "__auto_type.c:60:29: error: cannot use '__auto_type' with initializer list" \
+

